### ISSUE TITLE
Objects created by userFactory can access sharePublicMessage method

### ISSUE THREAD

**Issue by abhaasgoyal** [CodesmithLLC/cs-bin-solutions#4]
https://github.com/CodesmithLLC/cs-bin-solutions/blob/20de7d48a6207121afe528994418f7d769159875/oop.js#L181-L188

Above issue is with respect to Extension Question for - http://csbin.io/oop
The code does not prevent userFactory to access sharePublicMessage method.. Instead the solution should be to use Object.create in adminFactory function: 

```javascript
function adminFactory(name, score) {
  // Put code here
  let newAdminObj = userFactory(name,score);
  newAdminObj = Object.create(adminFunctionStore);
  newAdminObj.type = "Admin";
  return newAdminObj;
}

// Put code here for a method called sharePublicMessage

adminFunctionStore.sharePublicMessage = () => console.log("Welcome users!")
```

**Comment by samskito** [comment]
@abhaasgoyal You are right saying that the `userFactory` have access to `sharePublicMessage`. But your solution is incorrect, you override the `newAdminObj`. This new object now doesn't have the properties of `userFactory` (Both name and score will be undefined).

```javascript
// This part is right:
adminFunctionStore.sharePublicMessage = () => console.log("Welcome users!") 
```

Correct me if I'm wrong here, but the correct way should be to have both prototypes accessible to the object returned by the `adminFactory`. The way I did it was to create an object with the admin prototype, merge them and assign the new object as the new prototype of the returned object. This way we have all the methods in both function stores and the properties defined in `userFactory`.

```javascript
function adminFactory(name, score) {
  const admin = new userFactory(name, score)
  const withAdminMethods = Object.create(adminFunctionStore)
  Object.assign(Object.getPrototypeOf(withAdminMethods), Object.getPrototypeOf(admin))
  Object.setPrototypeOf(admin, withAdminMethods)
  admin.type = 'Admin'
  return admin
}
```

It is not an ideal solution but it works.

Intersting links:
- [Object.assign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)
- [Object.setPrototypeOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf)
